staging/src/k8s.io/client-go/tools/watch/retrywatcher.go

```diff
@@ -0,0 +1,232 @@
+	// Writing to an unbuffered channel is blocking operation
+	// and we need to check if stop wasn't requested while doing so.
+	select {
+	case rw.resultChan <- event:
+		return true
+	case <-rw.stopChan:
+		return false
+	}
+}
+
+// doReceive returns true when it is done, false otherwise.
+// If it is not done the second return value holds the time to wait before calling it again.
+func (rw *RetryWatcher) doReceive() (bool, time.Duration) {
+	watcher, err := rw.watcherFunc(rw.lastResourceVersion)
+	if err != nil {
+		klog.Errorf("retryWatcher: watcherFunc failed: %v", err)
+		// Retry
+		return false, 0
+	}
+
+	if watcher == nil {
+		klog.Error("retryWatcher: watcherFunc returned nil watcher")
+		// Retry
+		return false, 0
+	}
+
+	ch := watcher.ResultChan()
+	defer watcher.Stop()
+
+	for {
+		select {
+		case <-rw.stopChan:
+			klog.V(4).Info("Stopping RetryWatcher.")
+			return true, 0
+		case event, ok := <-ch:
+			if !ok {
+				klog.Warningf("RetryWatcher - getting event failed! Re-creating the watcher. Last RV: %s", rw.lastResourceVersion)
+				return false, 0
+			}
+
+			// We need to inspect the event and get ResourceVersion out of it
+			switch event.Type {
+			case watch.Added, watch.Modified, watch.Deleted:
+				metaObject, ok := event.Object.(resourceVersionGetter)
+				if !ok {
+					_ = rw.send(watch.Event{
+						Type:   watch.Error,
+						Object: &apierrors.NewInternalError(errors.New("retryWatcher: doesn't support resourceVersion")).ErrStatus,
+					})
+					// We have to abort here because this might cause lastResourceVersion inconsistency by skipping a potential RV with valid data!
+					return true, 0
+				}
+
+				resourceVersion := metaObject.GetResourceVersion()
+				if resourceVersion == "" {
+					_ = rw.send(watch.Event{
+						Type:   watch.Error,
+						Object: &apierrors.NewInternalError(fmt.Errorf("retryWatcher: object %#v doesn't support resourceVersion", event.Object)).ErrStatus,
+					})
+					// We have to abort here because this might cause lastResourceVersion inconsistency by skipping a potential RV with valid data!
+					return true, 0
+				}
+
+				// All is fine; send the event and update lastResourceVersion
+				ok = rw.send(event)
+				if !ok {
+					return true, 0
+				}
+				rw.lastResourceVersion = resourceVersion
+
+				continue
+
+			case watch.Error:
+				status, ok := event.Object.(*metav1.Status)
+				if !ok {
+					klog.Error(spew.Sprintf("RetryWatcher received an error which is not *metav1.Status but %#+v", event.Object))
+					// Retry unknown errors
+					return false, 0
+				}
+
+				switch status.Code {
+				case http.StatusGone:
+					// Never retry RV too old errors
+					_ = rw.send(event)
+					return true, 0
+
+				case http.StatusGatewayTimeout:
+					if status.Details == nil {
+						// Retry
+						return false, 0
+					}
+
+					// Retry after delay
+					return false, time.Duration(status.Details.RetryAfterSeconds) * time.Second
```

Comment:
is there a reason to limit this to StatusGatewayTimeout, and not to use `api/error#SuggestsClientDelay(err)` to check this?

Reply:
thx, I wasn't aware we had a function for it, will fix it

Reply:
I've made it common for all the failures but didn't use `SuggestsClientDelay` in the end as it requires  different type and creating another object embedding the type just to artificially get it din't feel right

Reply:
suggest extracting a helper method to check for delay directly on a status object in the api errors package like this:

```go
// SuggestsClientDelay returns true if this error suggests a client delay as well as the
// suggested seconds to wait, or false if the error does not imply a wait. It does not
// address whether the error *should* be retried, since some errors (like a 3xx) may
// request delay without retry.
func SuggestsClientDelay(err error) (int, bool) {
	if t, ok := err.(APIStatus); ok {
		return StatusSuggestsClientDelay(t.Status())
	}
	return 0, false
}

func StatusSuggestsClientDelay(status metav1.Status) (int, bool) {
	if status.Details == nil {
		return 0, false
	}
	if status.Reason == metav1.StatusReasonServerTimeout {
		// this StatusReason explicitly requests the caller to delay the action
		return int(status.Details.RetryAfterSeconds), true
	}
	if status.Details.RetryAfterSeconds > 0 {
		// If the client requests that we retry after a certain number of seconds
		return int(status.Details.RetryAfterSeconds), true
	}
	return 0, false
}
```

we really don't want to treat status codes and retryafter delays inconsistently

Reply:
hm, given we'd just ignore the bool value, that whole function comes down to: give me whatever value is in `RetryAfterSeconds`. 

If that's 0, retry immediately, wait the amount if not.

what Am I missing?

(I guess I can just extract the function anyways to move it forward, just don't see the value.)

Reply:
Also none of the usages of that function actually reads the int value (I wondered why we don't return time.Duration there) so it seems exclusively used to determine retry purposes given you can't infer it  being set from the int value.